autoscale-api/src/main/java/com/sequenceiq/periscope/client/AutoscaleClient.java

```diff
@@ -81,71 +74,63 @@ private AutoscaleClient(String autoscaleAddress, String identityServerAddress, S
-                accessToken = identityClient.getToken(user, password);
-            }
-            token = accessToken.getToken();
-            int exp = (int) (accessToken.getExpiresIn() * TOKEN_EXPIRATION_FACTOR);
-            LOGGER.info("Token has been renewed and expires in {} seconds", exp);
-            tokenCache.put(TOKEN_KEY, accessToken.getToken(), ExpirationPolicy.CREATED, exp, TimeUnit.SECONDS);
-            renewEndpoints(token);
-        }
+    public AutoScaleClusterV1Endpoint clusterEndpoint() {
+        return refreshIfNeededAndGet(AutoScaleClusterV1Endpoint.class);
     }
 
-    private void renewEndpoints(String token) {
-        MultivaluedMap<String, Object> headers = new MultivaluedHashMap<>();
-        headers.add("Authorization", "Bearer " + token);
-        t = client.target(autoscaleAddress).path(AutoscaleApi.API_ROOT_CONTEXT);
-        alertEndpoint = newResource(AlertEndpoint.class, headers);
-        autoScaleClusterV1Endpoint = newResource(AutoScaleClusterV1Endpoint.class, headers);
-        configurationEndpoint = newResource(ConfigurationEndpoint.class, headers);
-        historyEndpoint = newResource(HistoryEndpoint.class, headers);
-        policyEndpoint = newResource(PolicyEndpoint.class, headers);
-        LOGGER.info("Endpoints have been renewed for AutoscaleClient");
+    public ConfigurationEndpoint configurationEndpoint() {
+        return refreshIfNeededAndGet(ConfigurationEndpoint.class);
     }
 
-    private <C> C newResource(Class<C> resourceInterface, MultivaluedMap<String, Object> headers) {
-        return WebResourceFactory.newResource(resourceInterface, t, false, headers, Collections.emptyList(), EMPTY_FORM);
+    public HistoryEndpoint historyEndpoint() {
+        return refreshIfNeededAndGet(HistoryEndpoint.class);
     }
 
-    public AlertEndpoint alertEndpoint() {
-        refresh();
-        return alertEndpoint;
+    public PolicyEndpoint policyEndpoint() {
+        return refreshIfNeededAndGet(PolicyEndpoint.class);
     }
 
-    public AutoScaleClusterV1Endpoint clusterEndpoint() {
-        refresh();
-        return autoScaleClusterV1Endpoint;
+    private ExpiringMap<String, String> configTokenCache() {
+        return ExpiringMap.builder().variableExpiration().expirationPolicy(ExpirationPolicy.CREATED).build();
     }
 
-    public ConfigurationEndpoint configurationEndpoint() {
-        refresh();
-        return configurationEndpoint;
+    private synchronized <T> T refreshIfNeededAndGet(Class<T> clazz) {
+        String token = tokenCache.get(TOKEN_KEY);
+        if (token == null || endpointHolder == null) {
+            AccessToken accessToken;
+            accessToken = secret != null ? identityClient.getToken(secret) : identityClient.getToken(user, password);
+            token = accessToken.getToken();
+            int exp = (int) (accessToken.getExpiresIn() * TOKEN_EXPIRATION_FACTOR);
+            logger.info("Token has been renewed and expires in {} seconds", exp);
```

Comment:
hi @mhmxs I think we using in most of the class big capital LOGGER can use here as well that format?

Reply:
Negative, because in all other place logger is a static field. But here is an instance property, because it is used in a synchronized block